numpy/core/src/umath/cpuid.c

```diff
@@ -1,97 +0,0 @@
-#define _UMATHMODULE
-#define _MULTIARRAYMODULE
-#define NPY_NO_DEPRECATED_API NPY_API_VERSION
-
-#include <Python.h>
-
-#include "npy_config.h"
-
-#include "cpuid.h"
-
-#define XCR_XFEATURE_ENABLED_MASK 0x0
-#define XSTATE_SSE 0x2
-#define XSTATE_YMM 0x4
-#define XSTATE_ZMM 0x70
-
-/*
- * verify the OS supports avx instructions
- * it can be disabled in some OS, e.g. with the nosavex boot option of linux
- */
-static NPY_INLINE
-int os_avx_support(void)
-{
-#if HAVE_XGETBV
-    /*
-     * use bytes for xgetbv to avoid issues with compiler not knowing the
-     * instruction
-     */
-    unsigned int eax, edx;
-    unsigned int ecx = XCR_XFEATURE_ENABLED_MASK;
-    __asm__("xgetbv" : "=a" (eax), "=d" (edx) : "c" (ecx));
-    return (eax & (XSTATE_SSE | XSTATE_YMM)) == (XSTATE_SSE | XSTATE_YMM);
-#else
-    return 0;
-#endif
-}
-
-static NPY_INLINE
-int os_avx512_support(void)
-{
-#if HAVE_XGETBV
-    unsigned int eax, edx;
-    unsigned int ecx = XCR_XFEATURE_ENABLED_MASK;
-    unsigned int xcr0 = XSTATE_ZMM | XSTATE_YMM | XSTATE_SSE;
-    __asm__("xgetbv" : "=a" (eax), "=d" (edx) : "c" (ecx));
-    return (eax & xcr0) == xcr0;
-#else
-    return 0;
-#endif
-}
-
-static NPY_INLINE
-int cpu_supports_fma(void)
-{
-#ifdef __x86_64__
-    unsigned int feature = 0x01;
-    unsigned int a, b, c, d;
-    __asm__ volatile (
-        "cpuid"				"\n\t"
-	: "=a" (a), "=b" (b), "=c" (c), "=d" (d)
-	: "a" (feature));
-    /*
-     * FMA is the 12th bit of ECX
-     */
-    return (c >> 12) & 1;
-#else
-    return 0;
-#endif
-}
-
-/*
- * Primitive cpu feature detect function
- * Currently only supports checking for avx on gcc compatible compilers.
- */
-NPY_NO_EXPORT int
-npy_cpu_supports(const char * feature)
-{
-#ifdef HAVE___BUILTIN_CPU_SUPPORTS
-    if (strcmp(feature, "avx512f") == 0) {
-#ifdef HAVE___BUILTIN_CPU_SUPPORTS_AVX512F
-        return __builtin_cpu_supports("avx512f") && os_avx512_support();
-#else
-        return 0;
-#endif
-    }
-    else if (strcmp(feature, "fma") == 0) {
-        return cpu_supports_fma() && __builtin_cpu_supports("avx2") && os_avx_support();
```

Comment:
Why `FMA` require `AVX2` instead of `AVX`? any reason?

Reply:
Well I got it, the current  SIMD `FMA3` code requires `AVX2` but now I don't understand the point behind checking the availability of `FMA3` and `AVX2`. Is there any CPU that supports `AVX2` and at the same time doesn't support `FMA3`?!!!.

Reply:
> Where do we have any backward compatibility problems?

/cc @mattip, here

Reply:
For safety `FMA` is no longer alias for `AVX2` but its an independent feature detecting `AVX2` and `FMA3` together.

Reply:
> Is there any CPU that supports `AVX2` and at the same time doesn't support `FMA3`?!!!.

Example of a CPU with FMA3 and no AVX2: https://en.wikipedia.org/wiki/AMD_FX 
I know that's a discontinued product, but in general, if a feature has its own CPUID flag, I would think its safer to use them and not make assumptions.

Reply:
@r-devulap, I was meaning that all CPUs that support `AVX2` also supports `FMA3` but not the opposite, so checking only `AVX2` will be enough and that's what almost all projects do but after a deep search, I found new X86 chips re-designed by `VIA` support `AVX2` but without `FMA3` supports.
so currently I'm working on improving the infrastructure in gh-13516, it almost re-designed in order to be extremely flexible so it can fix this issue as follows:

- `AVX2` is no longer implies `FMA3` and both are separated  but in Intel compilers
  this behavior will change to force `FMA3` and `AVX2` to imply each other since there are no
  sperate flags for them only `-march=core-avx2` and `/arch:core-avx2`, same for features `AVX512F` and `AVX512CD`

- the new infrastructure configurable syntax must support multiple dispatch-able features in the same object, Example:
```` 
/*@targets FMA3 AVX2 (FMA3 AVX2) */ // parentheses() used for multiple targets
````
the above config will generate the follows:
- an object for `FMA3`, with requesting a runtime detect for `FMA3`
- an object for `AVX2`, with requesting a runtime detect for `AVX2`
- an object for `AVX2` and `FMA3`,  with requesting a runtime detect for `AVX2` and `FMA3`
the infrastructure will be dynamic enough to deal with the compiler limitations scenarios to avoid creating duplicated objects in case of intel compiler

I will update the gh-13516 soon with the new updates then I open it for review instead of the current draft status, also I want to add that I put all your previous comments into high considerations but some of them need to be replied within a code with a benchmark that proves the efficiency behind all this work. hopefully, at the end, I can make everyone get satisfied.